### ISSUE TITLE
Extension type `operator==` will mostly not work

### ISSUE THREAD

**Issue by davidmorgan** [dart-lang/macros#65]
Since the JSON types are extension types over `Map`, `operator==` will mostly not compare by value and will fail.

In https://github.com/dart-lang/macros/pull/60 I add an `equals` method that does work.

This is surprising and likely to trip people up.

Brainstorming:

Should we have a lint "don't use operator== on extension types"? If so it should presumably also catch them being used in `Set`s or as `Map` keys.

Another thing we could do, since we _usually_ control the underlying types, would be to make those implementations (e.g. `_TypedMap`, `_GrowableMap`) have deep equality / hash code. But that would be confusing if people use the `fromJson` contructors with standard maps.

**Comment by eernstg** [comment]
> Should we have a lint "don't use operator== on extension types"? If so it should presumably also catch them being used in `Set`s or as `Map` keys.

Makes sense! .. but we might add another requirement: Warn against using operator `==`

- when the static type of the receiver is an extension type, and
- when the behavior of `==` for the representation object differs from the desired behavior of `==` for the extension type.

For example, we might easily encounter situations where the representation object uses the operator `==` from `Object` (at least in the statically known type, and "as far as we know" in subtypes as well), and the same kind of equality is a good match for the extension type.

Perhaps the lint could detect metadata like `@equatable extension type ....`?
